### PR TITLE
Add document generation & math document

### PR DESCRIPTION
**Pull request by alesiong**
Tested on this (`document-gen`) branch, results:
https://cuhkshenzhen.github.io/CUHKSZLib/

Also optimize some ci option

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -33,8 +38,27 @@ compiler:
 script:
   - mkdir build && cd build
   - cmake -DTESTING=1 ..
-  - make
+  - make -j
```
Comment (Yexiaoxing) [inline .travis.yml]: -j4 or higher?

File: .travis.yml [hunk 2]
```diff
@@ -33,8 +38,27 @@ compiler:
...
   - CTEST_OUTPUT_ON_FAILURE=1 make test
 notifications:
   email: false
 sudo: false
+
+after_success: |
+  if  [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+    if [[ "$CXX" == "g++-4.9" ]]; then
```
Comment (Yexiaoxing) [inline .travis.yml]: I hate gcc (ignore this comment)

File: .travis.yml [hunk 3]
```diff
@@ -33,8 +38,27 @@ compiler:
...
+      cd $TRAVIS_BUILD_DIR
+      chmod +x generateDocument.sh
+      ./generateDocument.sh
+    fi
+  fi
+
+deploy:
+  provider: pages
+  skip_cleanup: true
+  github_token: $GITHUB_TOKEN # Set in travis-ci.org dashboard
```
Comment (Yexiaoxing) [inline .travis.yml]: whose token you have set?

**Comment by Yexiaoxing** [comment]
Reply to my comment and you can merge :)

File: .travis.yml (hunk 1 above)
Comment (alesiong) [inline .travis.yml]: If left blank, `make` will use as many jobs as possible

File: .travis.yml (hunk 3 above)
Comment (alesiong) [inline .travis.yml]: currently my token

**Pull request merged**